### ISSUE TITLE
Problem setting default values with TypoScript (heigth, width, ...)

### ISSUE THREAD

**Issue by sypets** [bobosch/ods_osm#109]
I am using the latest master with TYPO3 v11.

We try to give reasonable defaults on our site so the map will look good even if the editor did not make any changes in the Flexform. 

This used to work quite well, but now I see a problem because some more defaults are set in flexform and these will now override the TypoScript - even if you have not changed it in flexform! (and you can't change the defaults in the flexform).

## possible solution

* revert setting the default in the Flexform for these values

(It is unfortunate, that it is not possible to handle this better in TYPO3, e.g. set the default in Flexform to the value in TypoScript.)

## More details

This concerns:

* heigth
* width
* probably more ...

Example, **new** Flexform:

```
<width>
  <TCEforms>                                                      
    <label>LLL:EXT:ods_osm/Resources/Private/Language/locallang_db.xlf:tt_content.pi_flexform.width</label>
      <config>
         <type>input</type>
         <size>5</size>
         <default>80vw</default>
      </config>
    </TCEforms>
</width>
```

Previous:

```
<width>
  <TCEforms>
    <label>LLL:EXT:ods_osm/Resources/Private/Language/locallang_db.xlf:tt_content.pi_flexform.width</label>
    <config>
      <type>input</type>
      <size>5</size>
    </config>
  </TCEforms>
</width>
```

## Futher resources

In addition, I think there are also some problems in TYPO3 (which we can't solve right now), but perhaps get some inspiration and insight from the issues:

* you can't dynamically set the default for a flexform, e.g. with a user function, see https://stackoverflow.com/questions/56646328/typo3-tca-make-the-default-value-dynamic (there is also an issue which I can't find right now)
* combine Flexform / Typoscript issue: https://forge.typo3.org/issues/92363
* solution in news: https://docs.typo3.org/p/georgringer/news/8.4/en-us/AdministratorManual/Configuration/TypoScript/Index.html#overrideflexformsettingsifempty

**Comment by sypets** [comment]
@albig Apart from removing the default values again I do not have a good idea right now. Sorry.

**Comment by albig** [comment]
Please have a look at #112. I'm not fully sure about a solution, which fits all.

**Comment by sypets** [comment]
@albig I think we will try to fix the height and width problem on our end (so it looks ok with your defaults). Right now there is something weird which results in the maps not having the correct size and overlapping out of the div. Should figure this out first.  So when this is fixed the worst that could happen is the maps having an 80% width (and editors not really knowing what "80vh" is). (we used "100%" for width previously)

I am just not sure if there are other values which now have a default and didn't before.

#112 sounds good (though I am not sure how it was before, it sounds like the previous behaviour,).

> If the flexform setting is empty, the value of TypoScript will be used.

That is good, but that will not fix this issue where the behaviour was changed for the height and width to not be empty by default. For the TypoScript settings to apply, the editors would have to explicitly set it to empty (which is not so intuitive). 

And you are right: 

> If TypoScript should always overwrite Flexform, we must switch step 2 and 3. But in this case, it will be hard to understand for editors, why the changes in the Flexform have no effect.

I don't think TypoScript should always overwrite Flexform. In our case, that was not the intention. It was the intention to fall back to the TypoScript value **if a value is not given in Fleform**.

If there is a default value in height and width it is not possible to differentiate between "nothing was changed in Flexform" and "intentionally setting the value to the default value".

Again, I don't have a good solution without making it more complicated or reverting to previous without the (additional) defaults. So, don't worry about it. We will try to find a solution here.

**Comment by albig** [comment]
I'm not sure, if this was working with version 4.0.0. There are TypoScript settings which seems to have no effect.

From which version of ods_osm are you currently updating?

Currently I think this issue is indeed a bug. And I'll provide a solution. Not today but in the next days. After this we should be ready for release.

Am 2. Dezember 2022 10:07:43 MEZ schrieb Sybille Peters ***@***.***>:
>@albig I think we will try to fix the height and width problem on our end (so it looks ok with your defaults). Right now there is something weird which results in the maps not having the correct size and overlapping out of the div. Should figure this out first.  So when this is fixed the worst that could happen is the maps having an 80% width (and editors not really knowing what "80vh" is). (we used "100%" for width previously)
>
>I am just not sure if there are other values which now have a default and didn't before.
>
>#112 sounds good (though I am not sure how it was before, it sounds like the previous behaviour,).
>
>> If the flexform setting is empty, the value of TypoScript will be used.
>
>That is good, but that will not fix this issue where the behaviour was changed for the height and width to not be empty by default. For the TypoScript settings to apply, the editors would have to explicitly set it to empty (which is not so intuitive). 
>
>And you are right: 
>
>> If TypoScript should always overwrite Flexform, we must switch step 2 and 3. But in this case, it will be hard to understand for editors, why the changes in the Flexform have no effect.
>
>I don't think TypoScript should always overwrite Flexform. In our case, that was not the intention. It was the intention to fall back to the TypoScript value **if a value is not given in Fleform**.
>
>If there is a default value in height and width it is not possible to differentiate between "nothing was changed in Flexform" and "intentionally setting the value to the default value".
>
>Again, I don't have a good solution without making it more complicated or reverting to previous without the (additional) defaults. So, don't worry about it. We will try to find a solution here.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/bobosch/ods_osm/issues/109#issuecomment-1334956571
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- Von /e/ Mail gesendet.

**Comment by sypets** [comment]
Thanks, looking forward to it.

I had downgraded again to "[efa197d8d07c99b06d616fdd4f97d8bdd3d119ae](https://github.com/bobosch/ods_osm/commit/efa197d8d07c99b06d616fdd4f97d8bdd3d119ae)". This is not an actual release. I am not sure why I ended up exactly there, but there were problems after upgrading to 4.0.0 (probably due to the problem with the defaults and not all Flexforms contained values so they fell back to TypoScript with previous version.

From my end, we are ok with this version for now so there is no rush. Any future versions I can test first on a test system.
